src/runcrate/convert.py

```diff
@@ -18,793 +18,54 @@
-        manifest = {}
-        with open(self.root / Path(MANIFEST_FILE)) as f:
-            for line in f:
-                hash_, relpath = line.strip().split(None, 1)
-                manifest[hash_] = self.root / relpath
-        return manifest
-
-    def _resolve_plan(self, activity):
-        job_qname = activity.plan()
-        plan = activity.provenance.entity(job_qname)
-        if not plan:
-            m = SCATTER_JOB_PATTERN.match(str(job_qname))
-            if m:
-                plan = activity.provenance.entity(m.groups()[0])
-        return plan
-
-    def _get_hash(self, prov_param):
-        k = prov_param.id.localpart
-        try:
-            return self.hashes[k]
-        except KeyError:
-            type_names = frozenset(str(_) for _ in prov_param.types())
-            if "wf4ever:File" in type_names:
-                hash_ = next(prov_param.specializationOf()).id.localpart
-                self.hashes[k] = hash_
-                return hash_
-            elif "ro:Folder" in type_names:
-                m = hashlib.sha1()
-                m.update("".join(sorted(
-                    self._get_hash(_) for _ in self.get_dict(prov_param).values()
-                )).encode())
-                self.hashes[k] = hash_ = m.hexdigest()
-                return hash_
-
-    def _get_hashes(self, provenance):
-        for r in provenance.prov_doc.get_records(prov.model.ProvEntity):
-            self._get_hash(Entity(provenance, r))
-
-    def get_members(self, entity):
-        membership = entity.provenance.record_with_attr(
-            prov.model.ProvMembership, entity.id, prov.model.PROV_ATTR_COLLECTION
-        )
-        member_ids = (_.get_attribute(prov.model.PROV_ATTR_ENTITY) for _ in membership)
-        return (entity.provenance.entity(first(_)) for _ in member_ids)
-
-    def get_dict(self, entity):
-        d = {}
-        for qname in entity.record.get_attribute("prov:hadDictionaryMember"):
-            kvp = entity.provenance.entity(qname)
-            key = first(kvp.record.get_attribute("prov:pairKey"))
-            entity_id = first(kvp.record.get_attribute("prov:pairEntity"))
-            d[key] = entity.provenance.entity(entity_id)
-        return d
+    def __init__(self,
+                 root,
+                 converter=CONVERTERS["cwl"],
+                 workflow_name=None,
+                 license=None,
+                 readme=None):
+        self.converter = converter
+        self.converter.root = Path(root)
+        self.converter.workflow_name = workflow_name
+        self.converter.license = license
+        self.converter.readme = Path(readme) if readme else readme
+        self.converter.wf_path = self.converter.root / "workflow" / self.converter.WORKFLOW_BASENAME
+        self.converter.workflow_definition = self.converter.get_workflow()
+        self.converter.step_maps = self.converter.get_step_maps()
+        self.converter.ro = ResearchObject(BDBag(str(root)))
```

Comment:
This is CWLProv-specific. Other attributes that refer to `self.converter.ro` are also CWLProv-specific

Reply:
Yeah I was thinking I might move some of this into base and specialising bits into the derived classes